mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala

```diff
@@ -0,0 +1,220 @@
+      _._2
+    }, nClusters, nRuns)
+    vectRdd.unpersist()
+    if (logger.isDebugEnabled) {
+      logger.debug(s"Eigenvalue = $lambda EigenVector: ${localVt.mkString(",")}")
+    }
+    val estimates = vectRdd.zip(model.predict(vectRdd.map(_._2)))
+    if (logger.isDebugEnabled) {
+      logger.debug(s"lambda=$lambda  eigen=${localVt.mkString(",")}")
+    }
+    val ccs = (0 until model.clusterCenters.length).zip(model.clusterCenters)
+    if (logger.isDebugEnabled) {
+      logger.debug(s"Kmeans model cluster centers: ${ccs.mkString(",")}")
+    }
+    val estCollected = estimates.collect.sortBy(_._1._1)
+    if (logger.isDebugEnabled) {
+      val clusters = estCollected.map(_._2)
+      val counts = estCollected.groupBy(_._2).mapValues {
+        _.length
+      }
+      logger.debug(s"Cluster counts: Counts: ${counts.mkString(",")}"
+        + s"\nCluster Estimates: ${estCollected.mkString(",")}")
+    }
+    (ccs, estCollected)
+  }
+
+
+  /**
+   * Create a Graph given an initial Vt0 and a set of Edges that
+   * represent the Normalized Affinity Matrix (W)
+   */
+  def createGraphFromEdges(sc: SparkContext,
+                           edgesRdd: RDD[Edge[Double]],
+                           nPoints: Int,
+                           optInitialVt: Option[Seq[(VertexId, Double)]] = None) = {
+
+    assert(nPoints > 0, "Must provide number of points from the original dataset")
+    val G = if (optInitialVt.isDefined) {
+      val initialVt = optInitialVt.get
+      val vertsRdd = sc.parallelize(initialVt)
+      Graph(vertsRdd, edgesRdd)
+    } else {
+      Graph.fromEdges(edgesRdd, -1.0)
+    }
+    G
+
+  }
+
+  /**
+   * Calculate the dominant Eigenvalue and Eigenvector for a given sparse graph
+   * using the PIC method
+   * @param sc
+   * @param G  Input Graph representing the Normalized Affinity Matrix (W)
+   * @param nIterations Number of iterations of the PIC algorithm
+   * @param optMinNormChange Minimum norm acceleration for detecting convergence
+   *                         - indicated as "epsilon" in the PIC paper
+   * @return
+   */
+  def getPrincipalEigen(sc: SparkContext,
+                        G: DGraph,
+                        nIterations: Int = defaultIterations,
+                        optMinNormChange: Option[Double] = None
+                         ): (DGraph, Double, VertexRDD[Double]) = {
+
+    var priorNorm = Double.MaxValue
+    var norm = Double.MaxValue
+    var priorNormVelocity = Double.MaxValue
+    var normVelocity = Double.MaxValue
+    var normAccel = Double.MaxValue
+    val DummyVertexId = -99L
+    var vnorm: Double = -1.0
+    var outG: DGraph = null
+    var prevG: DGraph = G
+    val epsilon = optMinNormChange
+      .getOrElse(1e-5 / G.vertices.count())
+    for (iter <- 0 until nIterations
+         if Math.abs(normAccel) > epsilon) {
+
+      val tmpEigen = prevG.aggregateMessages[Double](ctx => {
+        ctx.sendToSrc(ctx.attr * ctx.srcAttr);
```

Comment:
This depends on how node attribute is defined. See my previous comment on documenting `G`.